### PR TITLE
[CALCITE-5647] RelMdPopulationSize should use mq.getRowCount(rel) instead of rel.estimateRowCount(mq)

### PR DESCRIPTION
**Pull request by jduo**
(From: #3151)
This PR corrects a row count call to use the RelMetadataQuery instead of going directly to rel.estimateRowCount.

Using rel.estimateRowCount directly will behave incorrectly when using a cardinality estimator that is implemented entirely in a MetadataHandler class, as it will evade the metadata handler's implementation of getRowCount and fall through to the underlying default method for Values.

This corrects the problem by using mq.getRowCount instead of rel.estimateRowCount. In the default case this results in the same value but in the custom cardinality model case the expected handler is called.

### CODE REVIEW

**Comment by jduo** [comment]
@libenchao @chunweilei

**Comment by mihaibudiu** [comment]
There are some style violations

**Comment by jduo** [comment]
> There are some style violations

Thanks. Should be correct now.

**Comment by JiajunBernoulli** [comment]
@jduo Thanks for your PR.

Just minor suggestion:
1. Commit message should start with `[CALCITE-5647]`.
2. We like adding Java Document and JIRA link before unit test.

File: core/src/test/java/org/apache/calcite/rel/metadata/RelMdPopulationSizeTest.java
Comment (JiajunBernoulli) [inline core/src/test/java/org/apache/calcite/rel/metadata/RelMdPopulationSizeTest.java:48]: Can we add it in `RelMetadataTest`?

**Comment by jduo** [comment]
Thanks @JiajunBernoulli.

Do you mean I should squash my commits and give it the [CALCITE-5647] prefix?

**Comment by JiajunBernoulli** [comment]
@jduo Yes, you can `squash` after any committer approve.

But now we maybe need some new commits to make PR better.

File: core/src/test/java/org/apache/calcite/rel/metadata/RelMdPopulationSizeTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/rel/metadata/RelMdPopulationSizeTest.java]: Done.

**Comment by jduo** [comment]
> @jduo Thanks for your PR.
> 
> Just minor suggestion:
> 
> 1. Commit message should start with `[CALCITE-5647]`.
> 2. We like adding Java Document and JIRA link before unit test.

Done.

**Comment by jduo** [comment]
@JiajunBernoulli @mihaibudiu , were there any other concerns to address in this PR?

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3661]: I am not sure that these 3 comments are useful.
In fact, this one confused me.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3650]: This isn't really displaying a bug that is being fixed.
From the JIRA comments I understand that it's difficult to create a small reproduction.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3661]: This is a pretty common way of documenting tests:
https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/

- Arrange would be setting up the test (that isn't global setup like the @Setup function)
- Act would be executing the test behavior
- Assert would be validating the result.

I can remove these comments but don't really see the harm in them.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3650]: The bug is that the population size was being calculated without utilizing RelMetadataQuery#getRowCount(). The idea was that this test demonstrates the getRowCount() call is being used.

Does that clarify the bug and purpose of the test? This sort of test you'd probably want to spy() instead but I didn't want to introduce Mockito into the core sub-module just for this.

**Comment by jduo** [comment]
Hi @mihaibudiu and @JiajunBernoulli , just wondering what the next steps are with this PR.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3661]: Comments should provide information that is not clear from the code. "Assert" for example does not. And "act" could mean many thing. I think it would be better without them.

But here the test is not showing that you fixed a bug, it is showing just that a specific method is called. The bug should manifest itself in producing an incorrect or suboptimal result for some code transformation. That would be a convincing test.

**Comment by jduo** [comment]
@mihaibudiu I've changed the test. It's still a bit artificial because it makes a RelMetadataQuery which intentionally reports an incorrect statistic but hopefully demonstrates how the bug is fixed. Hope the comments are clear enough.

**Comment by jduo** [comment]
Forgot to add @adamkennedy to this PR.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3649]: Yes, the test is still artificial.

**Review by mihaibudiu — APPROVED** [review]
I am approving, but I am still unsure about exactly what this improvement fixes.

**Comment by jduo** [comment]
> I am approving, but I am still unsure about exactly what this improvement fixes.

The intent with RelMetadataQuery is that the implementation can provide custom behavior for each of the statistics you can requet with it.

For the case of the population size of Values clauses, it should be based on the row count / 2. However RelMetadataQuery was not using invoking its (potentially custom) method to get the row count, it was just inlining the row count estimate from the node.

**Comment by mihaibudiu** [comment]
You are describing how something is implemented. A bug description should be something like "Query X produces result Y instead of the expected Z", or "Optimizer produces the plan X instead of the expected Y because it uses the wrong statistics".

**Comment by jduo** [comment]
> You are describing how something is implemented. A bug description should be something like "Query X produces result Y instead of the expected Z", or "Optimizer produces the plan X instead of the expected Y because it uses the wrong statistics".

I guess in this case, the original implementation happens to return the correct value, but is not behaving as a valid extension point. So I'm not sure what would be a good concrete test for this type of problem.

**Comment by mihaibudiu** [comment]
Supposedly you have changed this behavior because it causes some problems in some application someone cares about. The ideal test case would be a small reproduction of the bug which caused this quest in the first place. The JIRA cases suggests that a small reproduction is difficult.

**Comment by rubenada** [comment]
LGTM.
@jduo , Just to avoid confusion, would it be possible to align commit message with PR title and Jira title?

**Comment by jduo** [comment]
@rubenada , I've made the commit and PR have the same title. Should the JIRA have the same title now? It states the problem whereas the commit/PR describe what changed.

**Comment by rubenada** [comment]
@jduo, usually the commit message is the same as Jira subject (more info here: https://calcite.apache.org/develop/#contributing).

What if we use (for both Jira and commit) something that describes in-between the problem and the change, i.e.  `[CALCITE-5647] RelMdPopulationSize should use mq.getRowCount(rel) instead of rel.estimateRowCount(mq)`   ?

**Comment by jduo** [comment]
> @jduo, usually the commit message is the same as Jira subject (more info here: https://calcite.apache.org/develop/#contributing).
> 
> What if we use (for both Jira and commit) something that describes in-between the problem and the change, i.e. `[CALCITE-5647] RelMdPopulationSize should use mq.getRowCount(rel) instead of rel.estimateRowCount(mq)` ?

Sounds good. I've updated the name of the PR and the commit message. However I cannot seem to edit the JIRA title (perhaps because I'm not the original reporter of the issue).

**Comment by rubenada** [comment]
Thanks @jduo , I have updated the Jira title myself.

@mihaibudiu , @JiajunBernoulli  I think the PR is in a good shape, do you have further remarks? Otherwise I'll merge it in the next 24-48h

**Comment by mihaibudiu** [comment]
I don't have any comments.

**Pull request merged**
